File: app/src/main/java/dev/trubitsyn/lorforandroid/ui/topic/TopicFragment.java
/*
 * Copyright (C) 2015-2016 Nikola Trubitsyn
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package dev.trubitsyn.lorforandroid.ui.topic;

import android.content.Intent;
import android.os.Bundle;
import android.text.Html;
import android.text.method.LinkMovementMethod;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.core.widget.NestedScrollView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.engine.DiskCacheStrategy;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import dev.trubitsyn.lorforandroid.R;
import dev.trubitsyn.lorforandroid.api.ApiManager;
import dev.trubitsyn.lorforandroid.api.model.Topic;
import dev.trubitsyn.lorforandroid.api.model.Topics;
import dev.trubitsyn.lorforandroid.ui.ImageActivity;
import dev.trubitsyn.lorforandroid.ui.base.LoadableFragment;
import dev.trubitsyn.lorforandroid.util.DateUtils;
import dev.trubitsyn.lorforandroid.util.PreferenceUtils;
import dev.trubitsyn.lorforandroid.util.StringUtils;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class TopicFragment extends LoadableFragment {
    @BindView(R.id.topicScrollView) NestedScrollView scrollView;
    @BindView(R.id.topicTitle) TextView title;
    @BindView(R.id.topicTags) TextView tags;
    @BindView(R.id.topicAuthor) TextView author;
    @BindView(R.id.topicDate) TextView date;
    @BindView(R.id.topicImage) @Nullable ImageView image;
    @BindView(R.id.topicMessage) TextView message;
    private String url;
    private String imageUrl;
    private Topic topic;

    public static TopicFragment newInstance(String url, String imageUrl) {
        TopicFragment fragment = new TopicFragment();
        Bundle args = new Bundle();
        args.putString("url", url);
        args.putString("imageUrl", imageUrl);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(false);
        url = getArguments().getString("url");
        imageUrl = getArguments().getString("imageUrl");
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        View view = inflater.inflate(R.layout.fragment_topic, container, false);
        unbinder = ButterKnife.bind(this, view);
        return view;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        if (savedInstanceState == null) {
            fetchData();
        } else {
            stopRefreshAndShow();
            setTopic();
        }
    }

    @Override
    protected void fetchData() {
        if (StringUtils.isClub(url)) {
            showErrorView(R.string.error_access_denied);
        } else {
            Call<Topics> topics = ApiManager.INSTANCE.getApiTopic().getTopic(url);
            topics.enqueue(new Callback<Topics>() {
                @Override
                public void onResponse(Call<Topics> call, Response<Topics> response) {
                    if (response.body() != null) {
                        topic = response.body().topic;
                        stopRefreshAndShow();
                        setTopic();
                    } else {
                        showErrorView(R.string.error_network);
                    }
                }

                @Override
                public void onFailure(Call<Topics> call, Throwable t) {
                    showErrorView(R.string.error_network);
                }
            });
        }
    }

    private void setTopic() {
        title.setText(Html.fromHtml(topic.getTitle()));
        List<String> tagsList = topic.getTags();
        if (!tagsList.isEmpty()) {
            tags.setVisibility(View.VISIBLE);
            tags.setText(StringUtils.tagsFromStrings(tagsList));
        } else tags.setVisibility(View.GONE);

        if (image != null) {
            if (PreferenceUtils.shouldLoadImagesNow(getActivity())) {
                loadImageAndSetImageActivityListener();
            } else {
                image.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        loadImageAndSetImageActivityListener();
                    }
                });
            }
        }

        author.setText(topic.getAuthor().getNick());
        date.setText(DateUtils.getDate(topic.getPostDate()));
        message.setText(Html.fromHtml(topic.getMessage()));
        message.setMovementMethod(LinkMovementMethod.getInstance());
    }

    private void loadImageAndSetImageActivityListener() {
        assert (image) != null;
        image.setImageDrawable(null);
        Glide.with(getContext())
                .load(imageUrl)
                .diskCacheStrategy(DiskCacheStrategy.ALL)
                .into((image));

        image.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(getActivity(), ImageActivity.class);
                intent.putExtra("imageUrl", imageUrl);
                startActivity(intent);
            }
        });
    }

    @Override
    protected View dataView() {
        return scrollView;
    }
}

File: app/src/main/java/dev/trubitsyn/lorforandroid/ui/topic/TopicFragment.kt
/*
 * Copyright (C) 2015-2021 Nikola Trubitsyn (getsmp)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package dev.trubitsyn.lorforandroid.ui.topic

import android.os.Bundle
import android.text.method.LinkMovementMethod
import android.view.*
import android.widget.ImageView
import android.widget.TextView
import androidx.core.text.parseAsHtml
import androidx.core.widget.NestedScrollView
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.navigation.fragment.findNavController
import androidx.navigation.fragment.navArgs
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import dagger.hilt.android.AndroidEntryPoint
import dev.trubitsyn.lorforandroid.R
import dev.trubitsyn.lorforandroid.util.PreferenceUtils
import dev.trubitsyn.lorforandroid.util.StringUtils

@AndroidEntryPoint
class TopicFragment : Fragment() {
    private val scrollView by lazy { requireView().findViewById<NestedScrollView>(R.id.topicScrollView) }
    internal val title by lazy { requireView().findViewById<TextView>(R.id.topicTitle) }
    internal val tags by lazy { requireView().findViewById<TextView>(R.id.topicTags) }
    internal val author by lazy { requireView().findViewById<TextView>(R.id.topicAuthor) }
    internal val date by lazy { requireView().findViewById<TextView>(R.id.topicDate) }
    private val image by lazy { requireView().findViewById<ImageView>(R.id.topicImage) }
    private val message by lazy { requireView().findViewById<TextView>(R.id.topicMessage) }
    private val args by navArgs<TopicFragmentArgs>()
    private val viewModel by viewModels<TopicViewModel> { TopicViewModelFactory(args.url) }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(false)
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater.inflate(R.menu.menu_topic, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            android.R.id.home -> {
                findNavController().navigateUp()
                true
            }
            R.id.showComments -> {
                val action = TopicFragmentDirections.actionTopicToComment(args.url)
                findNavController().navigate(action)
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_topic, container, false)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        if (StringUtils.isClub(args.url)) {
            //showErrorView(R.string.error_access_denied)
            return
        }
        viewModel.getTopic().observe(viewLifecycleOwner, { topic ->
            //stopRefreshAndShow()
            setTopic(topic)
        })
    }

    private fun setTopic(topic: TopicItem) {
        title.text = topic.title.parseAsHtml()
        val tagsList = topic.tags
        if (tagsList.isNotEmpty()) {
            tags.visibility = View.VISIBLE
            tags.text = tagsList
        } else {
            tags.visibility = View.GONE
        }

        args.imageUrl?.let { imageUrl ->
            if (PreferenceUtils.shouldLoadImagesNow(requireContext())) {
                loadImageAndSetImageActivityListener(imageUrl)
            } else {
                image.setOnClickListener { loadImageAndSetImageActivityListener(imageUrl) }
            }
        }

        author.text = topic.author
        date.text = topic.postDate
        message.text = topic.message.parseAsHtml()
        message.movementMethod = LinkMovementMethod.getInstance()
    }

    private fun loadImageAndSetImageActivityListener(imageUrl: String) {
        image.setImageDrawable(null)
        Glide.with(requireContext())
                .load(imageUrl)
                .diskCacheStrategy(DiskCacheStrategy.ALL)
                .into(image)

        image.setOnClickListener {
            //val action = TopicFragmentDirections.actionTopicToImageActivity(bitmap)
            //findNavController().navigate(action)
        }
    }
}
